test/gtest/ucp/test_ucp_request.cc

```diff
@@ -100,3 +102,392 @@ UCS_TEST_P(test_ucp_request, test_request_query)
+    }
+
+    void init() override
+    {
+        if (!m_ucp_config->ctx.proto_enable) {
+            UCS_TEST_SKIP_R("reset is not supported for proto v1");
+        }
+
+        ucs::scoped_setenv snd_buf("UCX_TCP_SNDBUF", "8K");
+        ucs::scoped_setenv ib_queue_len("UCX_IB_TX_QUEUE_LEN", "65");
+        ucp_test::init();
+        connect();
+    }
+
+    void cleanup() override
+    {
+        for (auto &rkey : m_rkeys) {
+            ucp_rkey_destroy(rkey);
+        }
+
+        for (auto &memh : m_memhs) {
+            ASSERT_UCS_OK(ucp_mem_unmap(receiver().ucph(), memh));
+        }
+
+        ucp_test::cleanup();
+    }
+
+    void connect()
+    {
+        ucp_request_param_t param = {0};
+        param.op_attr_mask        = UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+        static const size_t size  = 1024;
+        std::vector<char> sbuf(size);
+        std::vector<char> rbuf(size);
+
+        sender().connect(&receiver(), get_ep_params());
+        receiver().connect(&sender(), get_ep_params());
+
+        void *pre_msg = ucp_tag_send_nbx(sender().ep(), sbuf.data(), size, 0,
+                                         &param);
+        void *pre_rcv = ucp_tag_recv_nbx(receiver().worker(), rbuf.data(), size,
+                                         0, 0, &param);
+        request_wait(pre_msg);
+        request_wait(pre_rcv);
+    }
+
+    void get_stream_data(std::vector<char> &rbuf)
+    {
+        size_t roffset            = 0;
+        const double timeout      = 10;
+        const ucs_time_t deadline = ucs::get_deadline(timeout);
+        ucs_status_ptr_t rdata;
+        size_t length;
+
+        do {
+            progress();
+            rdata = ucp_stream_recv_data_nb(receiver().ep(), &length);
+            if (rdata == NULL) {
+                continue;
+            }
+
+            memcpy(&rbuf[roffset], rdata, length);
+            roffset += length;
+            ucp_stream_data_release(receiver().ep(), rdata);
+        } while ((ucs_get_time() < deadline) && (roffset < rbuf.size()));
+    }
+
+    static ucs_status_t
+    am_data_cb(void *arg, const void *header, size_t header_length, void *data,
+               size_t length, const ucp_am_recv_param_t *param)
+    {
+        std::vector<char> *rbuf = (std::vector<char>*)arg;
+
+        memcpy((*rbuf).data(), data, length);
+        return UCS_OK;
+    }
+
+    void *send_am(std::vector<char> &sbuf, std::vector<char> *rbuf_p)
+    {
+        ucp_request_param_t req_param = {0};
+        static unsigned am_id         = 1;
+        ucp_am_handler_param_t param;
+        void *sreq;
+
+        param.field_mask = UCP_AM_HANDLER_PARAM_FIELD_ID |
+                           UCP_AM_HANDLER_PARAM_FIELD_CB |
+                           UCP_AM_HANDLER_PARAM_FIELD_ARG;
+        param.id         = am_id;
+        param.cb         = am_data_cb;
+        param.arg        = rbuf_p;
+
+        ucs_status_t status;
+        status = ucp_worker_set_am_recv_handler(receiver().worker(), &param);
```

Comment:
we need to set recv handler on receiver side only once during initialization

Reply:
need to have different handler per recv bufff